refactor!: deprecate sorting based on `apps.txt` in `get_installed_apps`

sagarvora opened the pull request:
## Issue

`frappe.get_hooks` currently returns value based on the sort order of `apps.txt`. This isn't good for multi-tenancy and doesn't really help since the `installed_apps` global should already be in a sequence. `apps.txt` also sometimes get overwritten by bench CLI.

## Proposed Change

Deprecate `sort` parameter. Always get installed apps in the order defined in `installed_apps` global.

BREAKING CHANGE: can break benches when `installed_apps` sequence is intended to be different from `apps.txt` sequence.

sagarvora committed:
refactor!: deprecate sorting based on `apps.txt` in `get_installed_apps`

sagarvora commented:
Not sure how to write a test for this 😅

sagarvora commented on frappe/utils/jinja.py:
```diff
@@ -109,8 +109,7 @@ def get_jloader():
 		if not apps:
-			apps = frappe.local.flags.web_pages_apps or frappe.get_installed_apps(sort=True)
-			apps.reverse()
+			apps = list(reversed(frappe.local.flags.web_pages_apps or frappe.get_installed_apps()))
```
This is being done instead of `apps.reverse` because we do NOT want to mutate the original object.

sagarvora committed:
fix: defer `local.all_apps` loading

sagarvora commented:
later, alligator.

SaiFi0102 commented:
@sagarvora Why did you close this?

sagarvora commented:
> @sagarvora Why did you close this?

Had a lot of unwanted side effects (see failing tests)

sagarvora commented:
> @sagarvora Why did you close this?

I don't recall, probably failing tests 😅 

You can take it up if you want (or I will take a stab at this later on when I get some time)

SaiFi0102 commented:
I need to figure it out as well .. but not sure how yet

ankush commented:
This largely seems okay. I dont get rationale behind using `apps.txt` and also it's basically random now except frappe being first. 


reference: 
1. https://github.com/frappe/frappe/issues/17762
2. https://github.com/frappe/bench/issues/1414

ankush committed:
refactor: filter out apps not installed on bench

ankush committed:
tests: clear DB transactions before all db calls

Because of repeatable read isolation,
changes from externally executed command dont reflect until transaction is ended.

barredterra commented:
@sagarvora does this also apply to translations? I noticed in the past that we were unable to overwrite translations from ERPNext with our custom app's `de.csv`.

sagarvora commented:
> @sagarvora does this also apply to translations? I noticed in the past that we were unable to overwrite translations from ERPNext with our custom app's `de.csv`.

No, the [current design](https://github.com/frappe/frappe/blob/433115f62d6c7a62b57ae74faebf9d3c92c4ec52/frappe/translate.py#L313) is that translations from apps are loaded in the sequence of `apps.txt`

Maybe loading translations from apps that aren't installed isn't very multi-tenancy friendly. If this is a problem, we can consider keeping separate caches for each site - but the memory cost will get multiplied by number of sites.

@ankush is working on retaining the order of `apps.txt` here: https://github.com/frappe/bench/pull/1415
